File: BSTIterator_LC173.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.Stack;

/**
 * In the constructor, do an inorder scan and store values in a queue so that next and hasNext become O(1)
 *
 * Time Complexity for constructor: O(N) to do a full inorder scan
 * Space Complexity: O(N) for storing inorder traversal and O(H) for recursion stack
 *
 * Time Complexity for next() and hasNext(): O(1)
 * Space Complexity for next() and hasNext(): O(1) although they are using stored Queue of size N
 */
class BSTIterator_LC173 {
    private Queue<Integer> minQ = new LinkedList<>();

    public BSTIterator_LC173(TreeNode root) {
        inorder(root);
    }

    public void inorder(TreeNode root) {
        Stack<TreeNode> s = new Stack<TreeNode>();
        TreeNode node = root;

        while (node != null || s.size() > 0) {
            while (node != null) {
                s.push(node);
                node = node.left;
            }
            node = s.pop();
            this.minQ.add(node.val);
            node = node.right;
        }
    }

    /**
     * @return the next smallest number
     */
    public int next() {
        if (!this.minQ.isEmpty())
            return this.minQ.poll();
        else
            return -1;
    }

    /**
     * @return whether we have a next smallest number
     */
    public boolean hasNext() {
        return !this.minQ.isEmpty();
    }
}

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 *
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator obj = new BSTIterator(root);
 * int param_1 = obj.next();
 * boolean param_2 = obj.hasNext();
 */

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator obj = new BSTIterator(root);
 * int param_1 = obj.next();
 * boolean param_2 = obj.hasNext();
 */

File: bst_iterator_lc173.py
"""
Name: Srinidhi Bhat 
Did it run on LC: Yes

Time Complexity: O(n) - inorder traversal will take O(n)
Space Complexity: O(n) - extra list used to store the lists

Logic: inorder traversal of BSt gives sorted order of the elements, store that
in an additional list

maintain a global pointer, so that when next() is called
we know which element to return

for hasnext, we can compare gloabal pointers current value with length of 
array.
"""
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class BSTIterator:
    def inorder(self,root):
        if root is None:
            return
        self.inorder(root.left)
        self.nodes_sorted.append(root.val)
        self.inorder(root.right)
        
        
    def __init__(self, root: TreeNode):
        self.nodes_sorted = []
        self.pointer = -1
        
        self.inorder(root)

    def next(self) -> int:
        """
        @return the next smallest number
        """
        self.pointer+=1
        
        return self.nodes_sorted[self.pointer]
        

    def hasNext(self) -> bool:
        """
        @return whether we have a next smallest number
        """
        return self.pointer+1 < len(self.nodes_sorted)
        


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()